blockchain/blocks/src/tipset.rs

```
@@ -246,12 +251,16 @@ where
             header.epoch() == first_header.epoch(),
             "epochs are not equal",
         )?;
-        verify(
-            header.miner_address() != first_header.miner_address(),
-            "miner_addresses are not distinct",
-        )?;
+
+        headers_len += 1;
+        headers_set.insert(*header.miner_address());
```

we could fail early by checking if insertion succeeded, given it's a set, no?